core/src/main/java/org/jdbi/v3/CollectorFactoryRegistry.java

```diff
@@ -50,40 +51,108 @@ void register(CollectorFactory factory) {
-            return getErasedType(type) == SortedSet.class
-                    ? Optional.of(Collectors.toCollection(TreeSet::new))
-                    : Optional.empty();
+        public Collector<?, ?, ?> build(Type containerType) {
+            return Collectors.toCollection(TreeSet::new);
         }
     }
 
-    private static class ListCollectorFactory<T> implements CollectorFactory {
+    private static class ListCollectorFactory implements CollectorFactory {
         @Override
-        public Optional<Collector<?, ?, ?>> build(Type type) {
-            return getErasedType(type) == List.class
-                    ? Optional.of(Collectors.toList())
-                    : Optional.empty();
+        public boolean accepts(Type containerType) {
+            return getErasedType(containerType) == List.class;
+        }
+
+        @Override
+        public Optional<Type> elementType(Type containerType) {
+            return findGenericParameter(containerType, List.class);
+        }
+
+        @Override
+        public Collector<?, ?, ?> build(Type containerType) {
+            return Collectors.toList();
         }
     }
 
-    private static class SetCollectorFactory<T> implements CollectorFactory {
+    private static class SetCollectorFactory implements CollectorFactory {
         @Override
-        public Optional<Collector<?, ?, ?>> build(Type type) {
-            return getErasedType(type) == Set.class
-                    ? Optional.of(Collectors.toSet())
-                    : Optional.empty();
+        public boolean accepts(Type containerType) {
+            return getErasedType(containerType) == Set.class;
+        }
+
+        @Override
+        public Optional<Type> elementType(Type containerType) {
+            return findGenericParameter(containerType, Set.class);
+        }
+
+        @Override
+        public Collector<?, ?, ?> build(Type containerType) {
+            return Collectors.toSet();
+        }
+    }
+
+    private static class OptionalCollectorFactory implements CollectorFactory {
+        @Override
+        public boolean accepts(Type containerType) {
+            return getErasedType(containerType) == Optional.class;
+        }
+
+        @Override
+        public Optional<Type> elementType(Type containerType) {
+            return findGenericParameter(containerType, Optional.class);
+        }
+
+        @Override
+        public Collector<?, ?, ?> build(Type containerType) {
+            return Collector.<Object, OptionalBuilder, Optional<Object>>of(
+                    OptionalBuilder::new,
+                    OptionalBuilder::set,
+                    (left, right) -> left.build().isPresent() ? left : right,
+                    OptionalBuilder::build);
+        }
+
+    }
+
+    private static class OptionalBuilder<T> {
+        private Optional<T> optional = Optional.empty();
+
+        public void set(T value) {
+            optional = Optional.of(value);
```

Comment:
Should this do something along the lines of `if (optional.isPresent()) throw new IllegalStateException();` ?